File: src/main/java/org/karel/karel/compiler/WallsMap.java
package org.karel.karel.compiler;

import java.util.HashSet;
import java.util.Objects;
import java.util.Set;

public class WallsMap {
    private final int rows;
    private final int cols;
    private final Set<Coordinates> walls;

    public WallsMap(int rows, int cols, Set<Coordinates> walls){
        this.rows = rows;
        this.cols = cols;
        this.walls = walls;
    }

    public WallsMap(int rows, int cols){
        this(rows, cols, new HashSet<>());
    }

    boolean hasWall(int x, int y, Direction direction){
        return switch (direction){
            case NORTH -> (y == rows) || walls.contains(new Coordinates(x, y, Direction.NORTH)) || walls.contains(new Coordinates(x, y+1, Direction.SOUTH));
            case EAST -> (x == cols) || walls.contains(new Coordinates(x, y, Direction.EAST)) || walls.contains(new Coordinates(x+1, y, Direction.WEST));
            case SOUTH -> (y == 1) || walls.contains(new Coordinates(x, y, Direction.SOUTH)) || walls.contains(new Coordinates(x, y-1, Direction.NORTH));
            case WEST -> (x == 1) || walls.contains(new Coordinates(x, y, Direction.WEST)) || walls.contains(new Coordinates(x-1, y, Direction.SOUTH));
        };
    }

    public int getRows() {
        return rows;
    }

    public int getCols() {
        return cols;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof WallsMap wallsMap)) return false;
        return rows == wallsMap.rows && cols == wallsMap.cols && Objects.equals(walls, wallsMap.walls);
    }

    @Override
    public int hashCode() {
        return Objects.hash(rows, cols, walls);
    }
}

File: src/main/resources/static/js/WallsMap.js
export class WallsMap {
    constructor(rows, cols, walls = new Set()) {
        this.rows = rows;
        this.cols = cols;
        this.walls = walls;
    }

    hasWall(x, y, direction) {
        switch (direction) {
            case 'NORTH':
                return y === this.rows || this.walls.has(`${x},${y},N`) || this.walls.has(`${x},${y+1},S`);
            case 'EAST':
                return x === this.cols || this.walls.has(`${x},${y},E`) || this.walls.has(`${x+1},${y},W`);
            case 'SOUTH':
                return y === 1 || this.walls.has(`${x},${y},S`) || this.walls.has(`${x},${y-1},N`);
            case 'WEST':
                return x === 1 || this.walls.has(`${x},${y},W`) || this.walls.has(`${x-1},${y},E`);
            default:
                return false;
        }
    }

    getRows() {
        return this.rows;
    }

    getCols() {
        return this.cols;
    }

    equals(other) {
        if (!(other instanceof WallsMap)) return false;
        return this.rows === other.rows && this.cols === other.cols && this.walls.size === other.walls.size && [...this.walls].every(wall => other.walls.has(wall));
    }

    hashCode() {
        return `${this.rows},${this.cols},${Array.from(this.walls).join('|')}`;
    }
}
